File: native/SundeeFundee/Sources/SundeeFundeeKit/DomainLayer/Benchmark/BenchmarkModels.swift
import Foundation

// MARK: - Benchmark Models

/// Intensity level of a benchmark (1-5)
public enum BenchmarkIntensity: Int, Codable, Sendable {
    case one = 1
    case two = 2
    case three = 3
    case four = 4
    case five = 5
}

/// How a benchmark is scored
public enum BenchmarkScoringType: String, Codable, Sendable {
    case time
    case roundsAndReps
    case load
    case reps
    case calories
    case distance
}

/// Readiness tier for attempting a benchmark
public enum ReadinessTier: String, Codable, Sendable {
    case peak
    case moderate
    case recovery
}

/// Category of benchmark
public enum BenchmarkCategory: String, Codable, Sendable, CaseIterable {
    case sundeeFundee = "Sundee Fundee"
    case classicWODs = "Classic WODs"
    case strength = "Strength"
    case endurance = "Endurance"
    case gymnastics = "Gymnastics"
    case generalFitness = "General Fitness"
}

/// A predefined or custom benchmark definition
public struct BenchmarkDefinition: Codable, Sendable, Identifiable {
    public let id: String
    public let name: String
    public let category: String
    public let workoutDescription: String
    public let scoringType: BenchmarkScoringType
    public let isPredefined: Bool
    public let sortOrder: Int
    public let intensity: BenchmarkIntensity?
    public let movementTags: [String]?
    public let equipment: [String]?
    public let timeDomain: String?
    public let coachNotes: String?

    public init(
        id: String,
        name: String,
        category: String,
        workoutDescription: String,
        scoringType: BenchmarkScoringType,
        isPredefined: Bool,
        sortOrder: Int,
        intensity: BenchmarkIntensity? = nil,
        movementTags: [String]? = nil,
        equipment: [String]? = nil,
        timeDomain: String? = nil,
        coachNotes: String? = nil
    ) {
        self.id = id
        self.name = name
        self.category = category
        self.workoutDescription = workoutDescription
        self.scoringType = scoringType
        self.isPredefined = isPredefined
        self.sortOrder = sortOrder
        self.intensity = intensity
        self.movementTags = movementTags
        self.equipment = equipment
        self.timeDomain = timeDomain
        self.coachNotes = coachNotes
    }
}

/// Result of a benchmark attempt
public struct BenchmarkResult: Codable, Sendable, Identifiable {
    public let id: String
    public let benchmarkId: String
    public let benchmarkName: String
    public let score: Double
    public let notes: String?
    public let date: Date
    public let cyclePhase: CyclePhase?

    public init(
        id: String,
        benchmarkId: String,
        benchmarkName: String,
        score: Double,
        notes: String? = nil,
        date: Date,
        cyclePhase: CyclePhase? = nil
    ) {
        self.id = id
        self.benchmarkId = benchmarkId
        self.benchmarkName = benchmarkName
        self.score = score
        self.notes = notes
        self.date = date
        self.cyclePhase = cyclePhase
    }
}

/// Readiness assessment for attempting a benchmark
public struct BenchmarkReadiness: Codable, Sendable {
    public let tier: ReadinessTier
    public let reason: String
    public let recommendation: String
    public let phaseMultiplier: Double
    public let emoji: String
    public let color: String

    public init(
        tier: ReadinessTier,
        reason: String,
        recommendation: String,
        phaseMultiplier: Double,
        emoji: String,
        color: String
    ) {
        self.tier = tier
        self.reason = reason
        self.recommendation = recommendation
        self.phaseMultiplier = phaseMultiplier
        self.emoji = emoji
        self.color = color
    }
}

/// Best attempt window based on cycle
public struct AttemptWindow: Codable, Sendable {
    public let startDate: Date
    public let endDate: Date
    public let reason: String

    public init(startDate: Date, endDate: Date, reason: String) {
        self.startDate = startDate
        self.endDate = endDate
        self.reason = reason
    }
}

File: android/core/src/main/java/com/sundeefundee/core/domain/benchmark/BenchmarkModels.kt
package com.sundeefundee.core.domain.benchmark

import com.sundeefundee.core.model.CyclePhase
import kotlinx.serialization.Serializable

// MARK: - Benchmark Intensity

/// Intensity level of a benchmark (1-5)
@Serializable
enum class BenchmarkIntensity(val value: Int) {
    ONE(1),
    TWO(2),
    THREE(3),
    FOUR(4),
    FIVE(5);
}

// MARK: - Benchmark Scoring Type

/// How a benchmark is scored
@Serializable
enum class BenchmarkScoringType {
    TIME,
    ROUNDS_AND_REPS,
    LOAD,
    REPS,
    CALORIES,
    DISTANCE;

    val displayName: String
        get() = when (this) {
            TIME -> "Time"
            ROUNDS_AND_REPS -> "Rounds + Reps"
            LOAD -> "Load"
            REPS -> "Reps"
            CALORIES -> "Calories"
            DISTANCE -> "Distance"
        }
}

// MARK: - Readiness Tier

/// Readiness tier for attempting a benchmark
@Serializable
enum class ReadinessTier {
    PEAK,
    MODERATE,
    RECOVERY;

    val displayName: String
        get() = when (this) {
            PEAK -> "Peak"
            MODERATE -> "Moderate"
            RECOVERY -> "Recovery"
        }
}

// MARK: - Benchmark Category

/// Category of benchmark
@Serializable
enum class BenchmarkCategory(val displayName: String) {
    SUNDEE_FUNDEE("Sundee Fundee"),
    CLASSIC_WODS("Classic WODs"),
    STRENGTH("Strength"),
    ENDURANCE("Endurance"),
    GYMNASTICS("Gymnastics"),
    GENERAL_FITNESS("General Fitness");
}

// MARK: - Benchmark Definition

/// A predefined or custom benchmark definition
@Serializable
data class BenchmarkDefinition(
    val id: String,
    val name: String,
    val category: String,
    val workoutDescription: String,
    val scoringType: BenchmarkScoringType,
    val isPredefined: Boolean,
    val sortOrder: Int,
    val intensity: BenchmarkIntensity? = null,
    val movementTags: List<String>? = null,
    val equipment: List<String>? = null,
    val timeDomain: String? = null,
    val coachNotes: String? = null
)

// MARK: - Benchmark Result

/// Result of a benchmark attempt
@Serializable
data class BenchmarkResult(
    val id: String,
    val benchmarkId: String,
    val benchmarkName: String,
    val score: Double,
    val notes: String? = null,
    val date: String, // ISO8601
    val cyclePhase: CyclePhase? = null
)

// MARK: - Benchmark Readiness

/// Readiness assessment for attempting a benchmark
@Serializable
data class BenchmarkReadiness(
    val tier: ReadinessTier,
    val reason: String,
    val recommendation: String,
    val phaseMultiplier: Double,
    val emoji: String,
    val color: String
)

// MARK: - Attempt Window

/// Best attempt window based on cycle
@Serializable
data class AttemptWindow(
    val startDate: String, // ISO8601
    val endDate: String, // ISO8601
    val reason: String
)
